### ISSUE TITLE
Response codes for index sealing

### ISSUE THREAD

**Issue by brwe** [elastic/elasticsearch#11251]
We still have to decide which response codes to return if index sealing failed. Currently we return PARTIAL_CONTENT(206) for partial success and CONFLICT(409) for complete failure but this is not really right as discussed here: https://github.com/elastic/elasticsearch/pull/11179#discussion_r30437649
Sorry @nik9000 with all the excitement I actually forgot to fix that before merging.

**Comment by s1monw** [comment]
++ thanks britta

**Comment by clintongormley** [comment]
PARTIAL_CONTENT isn't used anywhere else yet, but I'm wondering if we should, eg a search request where some shards failed?

CONFLICT we only use on version conflicts atm, which i kinda like, but I agree with @nik9000 that the description indicates that it is the best fit for this condition.  That said, wondering if we should just go for a plain 400 here?

**Comment by bleskes** [comment]
The tricky part is that this is a multi-index operation which make it's tricky to decide what the top level rest response should be. 

400 is bad, because the spec explicitly says:

```
The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.
```

especially the later part is simply not true.

Since this is all a best effort api, I would vote to keep things simple and return 200 if nothing exploded on the top level and keep the more intricate "couldn't do  it, pending operation" or some other shard level failure to be apparent from the response. See #11336 for proposed improved, cleaner (imho) response where there is a single counter to watch for (`_shards.failed`) which is similar to flush.

**Comment by nik9000** [comment]
Why is w3.org down when I need it?

> 400 is bad

+1

I don't like 200 because it means:
```
The request has succeeded.
```

But it didn't. I mean, I asked for a seal (or a synced flush or whatever) and it only partially worked. My first instinct with the seal command is to stick it in a ```do {response = seal()} while (response != 200);``` and I suspect that is what other folks will think too. I figure if a user is thinking deeply enough about the process then they can do the json parsing and figure out the response - but most users should just spin until they get a good seal.

As to the actual code - I don't care so long as it doesn't scream "wrong!" like 200 or 400 do to me. ```CONFLICT``` is fine with me. Anyone who is ok with partial seals has the wherewithall to look at the result and can distinguish a total failure from a partial one without a different http code. And conflict implies retry-able to me.

It looks like webdav uses 207 for this kind of thing. It might be ok to ape it here somewhat. Or not. Sorry for being a pain about this.

**Comment by bleskes** [comment]
re reading the spec for 409 conflict, it may be actually what we want:

```
The request could not be completed due to a conflict with the current state of the resource. This code is only allowed in situations where it is expected that the user might be able to resolve the conflict and resubmit the request. The response body SHOULD include enough ...
```

99% of failed synced flush (sorry Nik :) ) will be because of concurrent indexing.

I think @brwe already said she's OK with it (correct me if wrong!) . @clintongormley I think you had some reservations, but I think this is were we gravitate. Any objections left?

> Sorry for being a pain about this.

Not at all - These are good discussions! we take good care to choose correctly here, so users in years to come will not be confused..

**Comment by nik9000** [comment]
> Any objections left?

None - 409 sounds like the best choice to me.

**Comment by clintongormley** [comment]
> Why is w3.org down when I need it?

There's quite a good page on this little website called wikipedia. you may have heard of it... :)

Yeah, I think 409 is best.  Only for complete failure? and 206 for partial failure? Or 409 for both?

**Comment by bleskes** [comment]
There are indeed subtleties here:
- a complete failure we be an exception, which will carry it's own rest status
- a  failure that relates to an entire shard group (like no assigned primary, or in flight operation) . This will be 409.
- a shard level failure (one replica failed to sync-flush for what ever reason). For this one I'm torn a bit but will go with 409 for simplicity to avoid 200 on error. 

Note that 206 is tricky to me because the spec blames it on the client:

```
The server is delivering only part of the resource (byte serving) due to a range header sent by the client. 
```

**Comment by nik9000** [comment]
> There's quite a good page on this little website called wikipedia. you may have heard of it... :)

I wanted to use a primary source! :)

> a complete failure we be an exception, which will carry it's own rest status

Sure. These are mostly 5xx level stuff I imagine and for the most part should stop any automated processes until a person looks at them.

> a failure that relates to an entire shard group (like no assigned primary, or in flight operation) . This will be 409.

+1

> a shard level failure (one replica failed to sync-flush for what ever reason). For this one I'm torn a bit but will go with 409 for simplicity to avoid 200 on error.

Thanks. I think 409 is best - if the user wants to get cute and say "some of my indexes can restore slow, I don't care" then they can issue the synced flush commands to the list of indexes they want - or they can parse the 409. I think, at least.

**Comment by clintongormley** [comment]
Closed by #11336
